### PR TITLE
Wip 369 doc to models

### PR DESCRIPTION
**Pull request by damienbfs**
Moved documentation to pydantic models.

Redoc only shows first part of docstring (until \f character), while sphinx doc the entire docstring (see pictures).

Also added core documentation to sphinx and option to build locally.

### CODE REVIEW

**Comment by PaulineMauryL** [comment]
## Coverage report


 <details><summary>Click to see where and how coverage changed</summary><table><thead>
  <tr><th>File</th><th>Statements</th><th>Missing</th><th>Coverage</th><th>Coverage<br>(new stmts)</th><th>Lines missing</th></tr>
</thead>
<tbody><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server/routes</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-99c0780dd8ed58717831157f6d4d562bb50c6fae37b794494b5815a2fdc57d59">routes_admin.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-99c0780dd8ed58717831157f6d4d562bb50c6fae37b794494b5815a2fdc57d59"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-99c0780dd8ed58717831157f6d4d562bb50c6fae37b794494b5815a2fdc57d59"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-99c0780dd8ed58717831157f6d4d562bb50c6fae37b794494b5815a2fdc57d59"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-99c0780dd8ed58717831157f6d4d562bb50c6fae37b794494b5815a2fdc57d59"></a></td><td></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-d88094fcb2b2b1f1caa40c3745c00ef392219864f6fe3c291e0b5ef9db0d72d9">routes_dp.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-d88094fcb2b2b1f1caa40c3745c00ef392219864f6fe3c291e0b5ef9db0d72d9"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-d88094fcb2b2b1f1caa40c3745c00ef392219864f6fe3c291e0b5ef9db0d72d9"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-d88094fcb2b2b1f1caa40c3745c00ef392219864f6fe3c291e0b5ef9db0d72d9"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-d88094fcb2b2b1f1caa40c3745c00ef392219864f6fe3c291e0b5ef9db0d72d9"></a></td><td></td><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server/tests</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-7f7d18385b324497d5ebcf51c3e0c0172158683cd7f916a75a261815c51f617c">test_api.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-7f7d18385b324497d5ebcf51c3e0c0172158683cd7f916a75a261815c51f617c"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-7f7d18385b324497d5ebcf51c3e0c0172158683cd7f916a75a261815c51f617c"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-7f7d18385b324497d5ebcf51c3e0c0172158683cd7f916a75a261815c51f617c"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-7f7d18385b324497d5ebcf51c3e0c0172158683cd7f916a75a261815c51f617c"></a></td><td></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-0ce03d082552382cfaeadeba39a2db7a04f14d000a7ee3ef20e205a0dc05a0f2">test_api_diffprivlib.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-0ce03d082552382cfaeadeba39a2db7a04f14d000a7ee3ef20e205a0dc05a0f2"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-0ce03d082552382cfaeadeba39a2db7a04f14d000a7ee3ef20e205a0dc05a0f2"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-0ce03d082552382cfaeadeba39a2db7a04f14d000a7ee3ef20e205a0dc05a0f2"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-0ce03d082552382cfaeadeba39a2db7a04f14d000a7ee3ef20e205a0dc05a0f2"></a></td><td></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-8773f83f90aaf6fea21b4fd0e73b6c070a1473bfdee1b4392d216be0e93813b1">test_api_smartnoise_synth.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-8773f83f90aaf6fea21b4fd0e73b6c070a1473bfdee1b4392d216be0e93813b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-8773f83f90aaf6fea21b4fd0e73b6c070a1473bfdee1b4392d216be0e93813b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-8773f83f90aaf6fea21b4fd0e73b6c070a1473bfdee1b4392d216be0e93813b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-8773f83f90aaf6fea21b4fd0e73b6c070a1473bfdee1b4392d216be0e93813b1"></a></td><td></td></tbody>
<tfoot>
<tr>
<td><b>Project Total</b></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/227/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td>&nbsp;</td>
</tr>
</tfoot>
</table>

<sub>

This report was generated by [python-coverage-comment-action](https://github.com/py-cov-action/python-coverage-comment-action)

</sub>
</details>

**Comment by damienbfs** [comment]
image

**Comment by damienbfs** [comment]
image

**Comment by damienbfs** [comment]
image

File: client/notebooks/Demo_Client_Notebook.ipynb [hunk 1]
```diff
@@ -147,10 +147,24 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 32,
+   "execution_count": 1,
```
Comment (PaulineMauryL) [inline client/notebooks/Demo_Client_Notebook.ipynb:150]: There is an import error at the beginning and then the of the cells are empty
https://github.com/dscc-admin-ch/lomas/blob/wip_369_doc_to_models/client/notebooks/Demo_Client_Notebook.ipynb

File: core/lomas_core/error_handler.py [hunk 2]
```diff
@@ -121,5 +133,14 @@ async def internal_server_exception_handler(
         LOG.info(f"InternalServerException raised: {exc.error_message}")
         return JSONResponse(
             status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
-            content={"InternalServerException": INTERNAL_SERVER_ERROR},
+            content=jsonable_encoder(InternalServerExceptionModel()),
         )
+
+
+# Server error responses for DP queries
+SERVER_QUERY_ERROR_RESPONSES: dict[int | str, dict[str, Any]] = {
```
Comment (PaulineMauryL) [inline core/lomas_core/error_handler.py:141]: What is the goal of this and ...ExceptionModel exceptions versus what was there before ?

File: core/lomas_core/models/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    library: DPLibraries
+    """The external library that caused the exception."""
+    message: str
+    """Exception error message.
+
+    For exceptions from libraries external to the lomas packages.
+    """
+
+
+class UnauthorizedAccessExceptionModel(LomasServerExceptionModel):
```
Comment (PaulineMauryL) [inline core/lomas_core/models/exceptions.py:49]: aren't ExternalLibraryException and UnauthorizedAccessException always related to a query ? (just trying to understand)

File: core/lomas_core/models/requests.py [hunk 4]
```diff
@@ -46,100 +63,199 @@ class DummyQueryModel(QueryModel):
...
     dummy_seed: int
+    """The seed to set at the start of the dummy dataset generation."""
 
 
 # SmartnoiseSQL
 # ----------------------------------------------------------------------------
 class SmartnoiseSQLRequestModel(LomasRequestModel):
     """Base input model for a smarnoise-sql request."""
 
+    model_config = ConfigDict(
```
Comment (PaulineMauryL) [inline core/lomas_core/models/requests.py:76]: nice ^^

File: server/lomas_server/tests/test_api_diffprivlib.py [hunk 5]
```diff
@@ -83,10 +88,12 @@ def test_imputation(diffprivlib_body, imputer_strategy):
             # Should not work unknow imputation strategy
             response = test_imputation(example_diffprivlib, "i_do_not_exist")
             assert response.status_code == status.HTTP_400_BAD_REQUEST
-            assert response.json() == {
-                "InvalidQueryException": ""
-                + "Imputation strategy i_do_not_exist not supported."
-            }
+            assert (
```
Comment (PaulineMauryL) [inline server/lomas_server/tests/test_api_diffprivlib.py:91]: That must have been very annoying.. thanks

File: core/lomas_core/error_handler.py (hunk 2 above)
Comment (damienbfs) [reply core/lomas_core/error_handler.py:141]: This allows to list the possible responses that an endpoint can return:
```
@router.post(
    "/dummy_diffprivlib_query",
    dependencies=[Depends(server_live)],
    response_model=QueryResponse,
    responses=SERVER_QUERY_ERROR_RESPONSES,
    tags=["USER_DUMMY"],
)
```

The format of the dict in responses requires pydantic models. Using these also has the benefit that we can show the output model in the redoc documentation, not only for the standard response, but also for exceptions.

File: client/notebooks/Demo_Client_Notebook.ipynb (hunk 1 above)
Comment (damienbfs) [reply client/notebooks/Demo_Client_Notebook.ipynb]: Fixed it, there were a few issues with wrong import statements.

File: docs/build_docs.py
Comment (LancelotMarti) [inline docs/build_docs.py:1]: top!

**Review by LancelotMarti — APPROVED** [review]
Very clean, thanks!

File: core/lomas_core/models/exceptions.py (hunk 3 above)
Comment (PaulineMauryL) [reply core/lomas_core/models/exceptions.py:49]: It is here: https://dscc-admin-ch.github.io/lomas-docs/client_errors.html
Except for InternalServerException, I think all exceptions are always directly related to queries.

**Pull request merged**